### DISCUSSION TITLE
Decoupling detekt Gradle task from tasks graph of 'gradle build'

### DISCUSSION THREAD

**Discussion by gavvvr** [detekt/detekt#4196] [Q&A] [answered]
I have an existing Spring Boot project and I decided to add detekt to my CI/CD pipeline to import warnings from detekt to Sonar.

The problem is that the presence of `io.gitlab.arturbosch.detekt` plugin adds `detekt` task to tasks graph for `build`:

<details>
<summary>build sequence</summary>

❯ gradle build --dry-run
Starting a Gradle Daemon, 1 incompatible and 1 stopped Daemons could not be reused, use --status for details
:compileKotlin SKIPPED
:compileJava SKIPPED
:processResources SKIPPED
:classes SKIPPED
:bootJarMainClassName SKIPPED
:bootJar SKIPPED
:inspectClassesForKotlinIC SKIPPED
:jar SKIPPED
:assemble SKIPPED
:detekt SKIPPED
:compileTestKotlin SKIPPED
:compileTestJava SKIPPED
:processTestResources SKIPPED
:testClasses SKIPPED
:test SKIPPED
:check SKIPPED
:build SKIPPED
</details>

Changing CI pipeline to explicitly exclude `detekt` like this `gradle clean build -x detekt` is not an acceptable option for me. I would expect a way to decouple `detekt` task from `gradle build`-task-graph and call `detekt` task only on demand (when Sonar analysis is being performed). Did not find such an option in the plugin itself. What I've also tried: 

```
// unfortunately this one won't allow me to call detekt task on demand explicitly:
gradle.startParameter.excludedTaskNames.addAll(listOf("detekt"))
```

It's probably easy, but there is a lack of my experience with Gradle.

---

UPD: 
Having no choice here looks so unusual for me (who worked more with Maven). In Maven it is often your responsibility to tie plugin's tasks to Maven phase you need.

Here is the temporary workaround to achieve my goal: 

```groovy
if (!("detekt" in gradle.startParameter.taskNames)) {
    gradle.startParameter.excludedTaskNames << "detekt"
}
```

Still interested if there is a more clean solution

**Comment by cortinico** [comment, marked as answer]
> Still interested if there is a more clean solution

Sadly, not. That's a limitation of Gradle. The option you have are:

* Excluding the task with `-x detekt` from the CLI
* Adding a `detekt.enabled = false` to your `build.gradle`. But this will disable the `detekt` task always, even when you invoke it alone.
* Use a conditional approach to exclude the task like [the one suggested here](https://stackoverflow.com/a/33910454/3765748)
```
gradle.taskGraph.whenReady {
    taskGraph ->
        if (taskGraph.hasTask(check)) {
            detekt.enabled = false
        }
}
```
which is really similar to what you did.

**Comment by gavvvr** [comment]
Based on what I know now, my final solution to the problem is:

```kotlin
if ("detekt" !in gradle.startParameter.taskNames) {
    tasks.detekt { enabled = false }
}
```
